From 9f0350271b12cb61d93e25af43ba6e6eedfcc304 Mon Sep 17 00:00:00 2001
From: Zoltan Fridrich <zfridric@redhat.com>
Date: Wed, 25 Mar 2026 19:43:33 +0100
Subject: [PATCH] pkcs11/p11_cipher: make AES keys ephemeral objects

Signed-off-by: Zoltan Fridrich <zfridric@redhat.com>
---
 lib/pkcs11/p11_cipher.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/lib/pkcs11/p11_cipher.c b/lib/pkcs11/p11_cipher.c
index 8837aa803..5e64364cb 100644
--- a/lib/pkcs11/p11_cipher.c
+++ b/lib/pkcs11/p11_cipher.c
@@ -183,7 +183,6 @@ static int aes_set_key(struct p11_cipher_ctx *ctx, const void *key,
 	CK_ATTRIBUTE attrs[] = { { CKA_CLASS, &attr_class, sizeof(attr_class) },
 				 { CKA_KEY_TYPE, &attr_key_type,
 				   sizeof(attr_key_type) },
-				 { CKA_TOKEN, &attr_true, sizeof(attr_true) },
 				 { CKA_ENCRYPT, &attr_true, sizeof(attr_true) },
 				 { CKA_DECRYPT, &attr_true, sizeof(attr_true) },
 				 { CKA_LABEL, label, sizeof(label) - 1 },
-- 
2.53.0

